Allow unknown task time in QueueResizingEsTPE

williamrandolph opened the pull request:
The afterExecute method previously asserted that a TimedRunnable task
must have a positive execution time. However, the code in TimedRunnable
returns a value of -1 when a task time is unknown. Here, we expand the
logic in the assertion to allow for that possibility, and we don't
update our task time average if the value is negative.

Fixes #41448

williamrandolph committed:
Allow unknown task time in QueueResizingEsTPE

The afterExecute method previously asserted that a TimedRunnable task
must have a positive execution time. However, the code in TimedRunnable
returns a value of -1 when a task time is unknown. Here, we expand the
logic in the assertion to allow for that possibility, and we don't
update our task time average if the value is negative.

Fixes #41448

elasticmachine commented:
Pinging @elastic/es-core-infra

williamrandolph commented:
@elasticmachine run elasticsearch-ci/1

jaymode commented on server/src/main/java/org/elasticsearch/common/util/concurrent/QueueResizingEsThreadPoolExecutor.java:
```diff
@@ -151,8 +151,12 @@ protected void afterExecute(Runnable r, Throwable t) {
         final long taskExecutionNanos = timedRunnable.getTotalExecutionNanos();
-        assert taskExecutionNanos >= 0 : "expected task to always take longer than 0 nanoseconds, got: " + taskExecutionNanos;
-        executionEWMA.addValue(taskExecutionNanos);
+        assert taskExecutionNanos >= 0 || taskExecutionNanos == -1 :
```
I am a bit paranoid about things. I'd prefer if we also tracked if an exception occurred with a boolean that gets updated in the `onRejection` and `onFailure` methods. Then we can say for certain that a taskExecutionNanos of -1 was really due to an exception.

jaymode reviewed (changes requested):
This looks pretty good. I think we can make a change to be more explicit though.

williamrandolph committed:
Add a failure flag to TimedRunnable

In order to be sure that a task has an execution time of -1 because of
a failure, I'm adding a failure flag boolean to the TimedRunnable class.
If execution time is negative for some other reason, an assertion will
fail.

jaymode reviewed (approved):
LGTM
